src/python/tod/noise.py

```diff
@@ -1,102 +1,147 @@
+        mixmatrix (dict): Mixing matrix describing how the PSDs should
+            be combined for detector noise.  If provided, must contain
+            entries for every detector, and every key specified for a
+            detector must be defined in `freqs` and `psds`.
+        indices (dict):  Integer index for every PSD, useful for
+            generating indepedendent and repeateable noise realizations.
+            If absent, runnign indices will be assigned and provided.
+
+    Attributes:
+        detectors (list): List of detector names
+        keys (list): List of PSD names
+
+    Raises:
+        KeyError: If `freqs`, `psds`, `mixmatrix` or `indices` do not
+            include all relevant entries.
+        ValueError: If vector lengths in `freqs` and `psds` do not match.
+
     """
 
-    def __init__(self, detectors=None, freqs=None, psds=None):
-        self._dets = []
+    def __init__(self, *, detectors, freqs, psds, mixmatrix=None, indices=None):
+        self.detectors = list(sorted(detectors))
+        if mixmatrix is None:
+            # Default diagonal mixing matrix
+            self.keys = self.detectors
+            self._mixmatrix = None
+        else:
+            # Assemble the list of keys needed for the specified detectors
+            keys = set()
+            self._mixmatrix = {}
+            for det in self.detectors:
+                self._mixmatrix[det] = {}
+                for key, weight in mixmatrix[det].items():
+                    keys.add(key)
+                    self._mixmatrix[det][key] = weight
+            self.keys = list(sorted(keys))
+        if indices is None:
+            self._indices = {}
+            for i, key in enumerate(self.keys):
+                self._indices[key] = i
+        else:
+            self._indices = dict(indices)
         self._freqs = {}
         self._psds = {}
-        self._weights = {}
-        self._rate = {}
-
-        if detectors is not None:
-            self._dets = detectors
-            if psds is None:
-                raise RutimeError("you must provide a dictionary of PSD arrays for all detectors")
-            if freqs is None:
-                raise RutimeError("you must provide a dictionary of frequency arrays for all detectors")
-            for det in self._dets:
-                if det not in psds:
-                    raise RuntimeError("no PSD specified for detector {}".format(det))
-                if det not in freqs:
-                    raise RuntimeError("no frequency array specified for detector {}".format(det))
-                if psds[det].shape[0] != freqs[det].shape[0]:
-                    raise RuntimeError("PSD length must match the number of frequencies")
-                self._freqs[det] = np.copy(freqs[det])
-                self._psds[det] = np.copy(psds[det])
-                # last frequency point should be Nyquist
-                self._rate[det] = 2.0 * self._freqs[det][-1]
-
-
-    @property
-    def detectors(self):
```

Comment:
I disagree.  The list of detectors was intended to be set at construction time and never changed by the user.  Although the user can of course manually change the internal _dets variable, the use of @property is fairly standard practice:

https://stackoverflow.com/questions/14594120/python-read-only-property